### PR TITLE
Build controller HA integration tests

### PR DESCRIPTION
**Pull request by csrwng**

### CODE REVIEW

**Comment by csrwng** [comment]
[test]

**Comment by csrwng** [comment]
Added long comment.
@bparees @smarterclayton ptal

File: pkg/build/controller/image_change_controller.go [hunk 1]
```diff
@@ -98,6 +104,21 @@ func (c *ImageChangeController) HandleImageRepo(repo *imageapi.ImageStream) erro
...
 
 		if shouldBuild {
+			// The following update is meant to reduce the chance that the image change controller
+			// will kick off multiple builds on an image change in a HA setup, where multiple controllers
+			// of the same type may be looking at the same etcd data.
+			// If multiple controllers read the same build config (with same ResourceVersion) above and
+			// make a determination that a build needs to be kicked off, the update will only allow one of
+			// those controllers to continue to launch the build, while the rest will return an error and
+			// reset their queue. This won't eliminate the chance of multiple builds, since another controller
+			// can read the build after this update and launch its own build.
```
Comment (bparees) [inline pkg/build/controller/image_change_controller.go]: +1

File: pkg/build/controller/image_change_controller_test.go [hunk 2]
```diff
@@ -240,7 +240,7 @@ func TestBuildConfigInstantiatorError(t *testing.T) {
 	if actual, expected := bcInstantiator.newBuild.Parameters.Strategy.DockerStrategy.From.Name, "registry.com/namespace/imagename:newImageID123"; actual != expected {
 		t.Errorf("Image substitutions not properly setup for new build. Expected %s, got %s |", expected, actual)
 	}
-	if bcUpdater.buildcfg != nil {
+	if bcUpdater.buildcfg != nil && bcUpdater.updateCount > 1 {
```
Comment (bparees) [inline pkg/build/controller/image_change_controller_test.go]: shouldn't this be 0 if we're expecting no updates?

Comment (csrwng) [inline pkg/build/controller/image_change_controller_test.go]: we are expecting the second update not to happen, but the first one does happen before we attempt to launch the build.

Comment (bparees) [inline pkg/build/controller/image_change_controller_test.go]: how does bcUpdater.updateCount get to 1 without bcUpdate.buildcfg getting set to non-nil?

Comment (csrwng) [inline pkg/build/controller/image_change_controller_test.go]: The test throws an error here:
https://github.com/csrwng/origin/blob/build_controller_ha/pkg/build/controller/image_change_controller.go#L125
which is after an update happened here:
https://github.com/openshift/origin/pull/2215/files#diff-717854e27d42bef56a972e64d148b529R116
After the update, buildcfg is non-nil

Comment (csrwng) [inline pkg/build/controller/image_change_controller_test.go]: @bparees I just realized I didn't answer what you were asking. There's no way for updateCount to be  1 without bldCfg being set to non-nil. Will remove the extraneous condition.

File: test/integration/buildcontroller_test.go [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+
+	// Start watching builds for New -> Pending transition
+	buildWatch, err := osClient.Builds(ns).Watch(labels.Everything(), fields.OneTermEqualSelector("name", b.Name), b.ResourceVersion)
+	checkErr(t, err)
+	defer buildWatch.Stop()
+	buildModifiedCount := int32(0)
+	go func() {
+		for e := range buildWatch.ResultChan() {
+			// Ignore any event that is not a modified event
+			if e.Type != watchapi.Modified {
```
Comment (bparees) [inline test/integration/buildcontroller_test.go]: hm. what other events come in?  shouldn't they only be modified events?

File: test/integration/buildcontroller_test.go [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+			defer buildWatch.Stop()
+			go func() {
+				done := false
+				for e := range buildWatch.ResultChan() {
+					var ok bool
+					b, ok = e.Object.(*buildapi.Build)
+					if !ok {
+						errChan <- fmt.Errorf("%s: unexpected object received: %#v", test.Name, e.Object)
+					}
+					if e.Type != watchapi.Modified {
```
Comment (bparees) [inline test/integration/buildcontroller_test.go]: again what other event types would be valid here?

File: test/integration/buildcontroller_test.go (hunk 3 above)
Comment (csrwng) [inline test/integration/buildcontroller_test.go]: no other type, actually. Will throw an error.

File: test/integration/buildcontroller_test.go (hunk 4 above)
Comment (csrwng) [inline test/integration/buildcontroller_test.go]: updated and throw an error as well.

**Comment by bparees** [comment]
lgtm.

**Comment by csrwng** [comment]
got bparee's lgtm... [merge]

**Pull request merged**